File: app/src/main/java/be/hcpl/android/filmtag/EditFrameFragment.java
package be.hcpl.android.filmtag;

import android.Manifest;
import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.content.pm.PackageManager;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.location.Location;
import android.location.LocationListener;
import android.location.LocationManager;
import android.net.Uri;
import android.os.Bundle;
import android.os.Environment;
import android.preference.PreferenceManager;
import android.provider.MediaStore;
import android.support.annotation.Nullable;
import android.support.v4.content.ContextCompat;
import android.text.TextUtils;
import android.util.Log;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.widget.CheckBox;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.TextView;
import android.widget.Toast;

import java.io.File;
import java.io.IOException;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Date;
import java.util.List;

import be.hcpl.android.filmtag.model.Frame;
import be.hcpl.android.filmtag.model.Roll;
import be.hcpl.android.filmtag.template.TemplateFragment;
import be.hcpl.android.filmtag.util.StorageUtil;
import butterknife.Bind;

/**
 * Created by hcpl on 1/08/15.
 */
public class EditFrameFragment extends TemplateFragment {

    // all image related code from http://developer.android.com/training/camera/photobasics.html

    private static final String KEY_FRAME_IDX = "frame_index";
    private static final String KEY_FRAMES = "frames";
    private static final String KEY_ROLL = "roll";
    private static final int MY_PERMISSIONS_REQUEST_LOCATION = 100;
    private static final int MY_PERMISSIONS_REQUEST_STORAGE = 200;
    private static final int REQUEST_IMAGE_CAPTURE = 1;

    @Bind(R.id.edit_shutter)
    EditText editShutter;
    @Bind(R.id.edit_aperture)
    EditText editAperture;
    @Bind(R.id.edit_notes)
    EditText editNotes;
    @Bind(R.id.long_exposure)
    CheckBox checkLongExposure;

    @Bind(R.id.image_preview)
    ImageView imagePreview;
    @Bind(R.id.image_preview_indicator)
    ImageView imagePreviewIndicator;
    @Bind(R.id.image_location_indicator)
    ImageView imageLocationIndicator;

    @Bind(R.id.edit_tags)
    EditText editTags;

    @Bind(R.id.text_location)
    TextView locationView;

    // model object of a film or roll with a number of frames exposed at a given value
    private Roll roll;

    private Frame selectedFrame;

    private List<Frame> frames;

    private static boolean locationPermissionRequested = false;
    private static boolean storagePermissionRequested = false;
    private static boolean storagePermissionRequestedForPreview = false;

    public static EditFrameFragment newInstance(Roll roll, List<Frame> frames, int frame) {
        Bundle args = new Bundle();
        args.putSerializable(KEY_FRAMES, (ArrayList) frames);
        args.putInt(KEY_FRAME_IDX, frame);
        args.putSerializable(KEY_ROLL, roll);
        EditFrameFragment fragment = new EditFrameFragment();
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    protected int getLayoutResourceId() {
        return R.layout.fragment_form_frame;
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setHasOptionsMenu(true);
        Bundle args = getArguments();
        if (args != null) {
            frames = (List<Frame>) args.getSerializable(KEY_FRAMES);
            selectedFrame = frames.get(args.getInt(KEY_FRAME_IDX));
            roll = (Roll) args.getSerializable(KEY_ROLL);
        }
    }

    @Override
    public void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);
        outState.putSerializable(KEY_FRAMES, (ArrayList) frames);
        outState.putInt(KEY_FRAME_IDX, frames.indexOf(selectedFrame));
        outState.putSerializable(KEY_ROLL, roll);
    }

    @Override
    public void onViewStateRestored(@Nullable Bundle savedInstanceState) {
        super.onViewStateRestored(savedInstanceState);
        if (savedInstanceState != null) {
            frames = (List<Frame>) savedInstanceState.getSerializable(KEY_FRAMES);
            selectedFrame = frames.get(savedInstanceState.getInt(KEY_FRAME_IDX));
            roll = (Roll) savedInstanceState.getSerializable(KEY_ROLL);
        }
    }

    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        inflater.inflate(R.menu.update_frame, menu);
    }

    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        if (selectedFrame != null) {
            ((TextView) view.findViewById(R.id.edit_number)).setText(String.valueOf(selectedFrame.getNumber()));
            if (selectedFrame.getAperture() != 0)
                editAperture.setText(String.valueOf(selectedFrame.getAperture()));
            if (selectedFrame.getShutter() != 0)
                editShutter.setText(String.valueOf(selectedFrame.getShutter()));
            checkLongExposure.setChecked(selectedFrame.isLongExposure());
            editNotes.setText(selectedFrame.getNotes());
            // populate the tags here
            if (selectedFrame.getTags() != null && !selectedFrame.getTags().isEmpty())
                editTags.setText(TextUtils.join(" ", selectedFrame.getTags()));
            loadImagePreview();
            showLocation();
        }

        SharedPreferences prefs = PreferenceManager.getDefaultSharedPreferences(getContext());
        editShutter.setHint(prefs.getString("key_default_shutter", String.valueOf(60)));
        editAperture.setHint(prefs.getString("key_default_apertures", String.valueOf(4)));

        // TODO implement autocomplete
    }

    private void markImageAvailable() {
        imagePreviewIndicator.setImageDrawable(selectedFrame.getPathToImage() != null ?
                ContextCompat.getDrawable(getActivity(), R.drawable.ic_action_image_photo_camera_primary) :
                ContextCompat.getDrawable(getActivity(), R.drawable.ic_action_image_photo_camera_silver));
    }

    private void markLocationAvailable() {
        imageLocationIndicator.setImageDrawable(selectedFrame.getLocation() != null ?
                ContextCompat.getDrawable(getActivity(), R.drawable.ic_action_device_gps_primary) :
                ContextCompat.getDrawable(getActivity(), R.drawable.ic_action_device_gps_silver));
        if (selectedFrame.getLocation() != null) {
            View.OnClickListener onClickListener = new View.OnClickListener() {
                @Override
                public void onClick(View view) {
                    showMap(Uri.parse("geo:" + selectedFrame.getLocation().getLatitude() + "," + selectedFrame.getLocation().getLongitude()));
                }
            };
            imageLocationIndicator.setOnClickListener(onClickListener);
            locationView.setOnClickListener(onClickListener);
        }
    }

    private void showMap(Uri geoLocation) {
        if (geoLocation == null)
            return;
        Intent intent = new Intent(Intent.ACTION_VIEW);
        intent.setData(geoLocation);
        if (intent.resolveActivity(getActivity().getPackageManager()) != null) {
            startActivity(intent);
        }
    }

    private void loadImagePreview() {
        // mark image available first so that if we don't have permission it's still marked
        markImageAvailable();
        // and if path set try loading after permission check
        if (selectedFrame.getPathToImage() != null) {
            // for this storage permission also required
            if (ContextCompat.checkSelfPermission(getActivity(),
                    Manifest.permission.WRITE_EXTERNAL_STORAGE) != PackageManager.PERMISSION_GRANTED) {
                if( !storagePermissionRequestedForPreview ) {
                    storagePermissionRequestedForPreview = true;
                    requestPermissions(new String[]{Manifest.permission.WRITE_EXTERNAL_STORAGE},
                            MY_PERMISSIONS_REQUEST_STORAGE);
                }
                return;
            }
            try {
                BitmapFactory.Options options = new BitmapFactory.Options();
                options.inSampleSize = 4;
                options.inJustDecodeBounds = false;
                Bitmap bm = BitmapFactory.decodeFile(selectedFrame.getPathToImage(), options);
                imagePreview.setImageBitmap(bm);
                // and try updating location
//                extractLocationFromFile();
            } catch (Exception e) {
                // ignore any exceptions here
                Log.e(getTag(), "failed to load image from configured path", e);

            }
        }
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.action_update:
                updateItem();
                return true;
            case android.R.id.home:
                backToOverview();
                return true;
            case R.id.action_camera:
                dispatchTakePictureIntent();
                return true;
            case R.id.action_location:
                getLocation();
                return true;
        }
        return false;
    }


    private void dispatchTakePictureIntent() {
        // check permissions first
        if (ContextCompat.checkSelfPermission(getActivity(),
                Manifest.permission.WRITE_EXTERNAL_STORAGE) != PackageManager.PERMISSION_GRANTED) {

            // TODO ignore rationale for now
//            if (ActivityCompat.shouldShowRequestPermissionRationale(getActivity(),
//                    Manifest.permission.ACCESS_FINE_LOCATION)) { //... } else {

            // No explanation needed, we can request the permission.
            if( !storagePermissionRequested ) {
                storagePermissionRequested = true;
                requestPermissions(
                        new String[]{Manifest.permission.WRITE_EXTERNAL_STORAGE},
                        MY_PERMISSIONS_REQUEST_STORAGE);
            }
            return;

        }

        Intent takePictureIntent = new Intent(MediaStore.ACTION_IMAGE_CAPTURE);
        // Ensure that there's a camera activity to handle the intent
        if (takePictureIntent.resolveActivity(getActivity().getPackageManager()) != null) {
            // Create the File where the photo should go
            File photoFile = null;
            try {
                photoFile = createImageFile();
            } catch (IOException ex) {
                // Error occurred while creating the File
                Log.e(getTag(), "failed to create temp image file", ex);
                Toast.makeText(getActivity(), R.string.error_creating_image, Toast.LENGTH_SHORT).show();
            }
            // Continue only if the File was successfully created
            if (photoFile != null) {
                takePictureIntent.putExtra(MediaStore.EXTRA_OUTPUT,
                        Uri.fromFile(photoFile));
                startActivityForResult(takePictureIntent, REQUEST_IMAGE_CAPTURE);
            }
        }
    }

    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        if (requestCode == REQUEST_IMAGE_CAPTURE && resultCode == Activity.RESULT_OK) {
//            Bundle extras = data.getExtras();
//            Bitmap imageBitmap = (Bitmap) extras.get("data");
            // show the image
//            imagePreview.setImageBitmap(imageBitmap);
            loadImagePreview();
        }
    }

    private File createImageFile() throws IOException {
        // Create an image file name
        String timeStamp = new SimpleDateFormat("yyyyMMdd_HHmmss").format(new Date());
        String imageFileName = "JPEG_" + timeStamp + "_";
        File storageDir = Environment.getExternalStoragePublicDirectory(
                Environment.DIRECTORY_PICTURES);
        File image = File.createTempFile(
                imageFileName,  /* prefix */
                ".jpg",         /* suffix */
                storageDir      /* directory */
        );

        // Save a file: path for use with ACTION_VIEW intents
        selectedFrame.setPathToImage(image.getAbsolutePath());
        return image;
    }

//    private void extractLocationFromFile() throws IOException {
//        ExifInterface exif = new ExifInterface(selectedFrame.getPathToImage());
//        String lat = exif.getAttribute(ExifInterface.TAG_GPS_LATITUDE);
//        String lon = exif.getAttribute(ExifInterface.TAG_GPS_LONGITUDE);
//        if (lat != null && lon != null) {
//            locationView.setText(getString(R.string.label_location) + " " + lat + " " + lon);
//        }
//    }

    private void getLocation() {
        if (ContextCompat.checkSelfPermission(getActivity(),
                Manifest.permission.ACCESS_FINE_LOCATION) != PackageManager.PERMISSION_GRANTED) {

            // TODO ignore rationale for now
//            if (ActivityCompat.shouldShowRequestPermissionRationale(getActivity(),
//                    Manifest.permission.ACCESS_FINE_LOCATION)) { //... } else {

            // No explanation needed, we can request the permission.
            if( !locationPermissionRequested ){
            locationPermissionRequested = true;
            requestPermissions(
                    new String[]{Manifest.permission.ACCESS_FINE_LOCATION},
                    MY_PERMISSIONS_REQUEST_LOCATION);
            }
            return;
        }
        registerLocationListener(LocationManager.GPS_PROVIDER);
    }

    /**
     * register location listeners
     *
     * @param provider
     */
    private void registerLocationListener(final String provider) {
        if (ContextCompat.checkSelfPermission(getActivity(),
                Manifest.permission.ACCESS_FINE_LOCATION) != PackageManager.PERMISSION_GRANTED) {
            return;
        }
        // remove previous listener first
        unregisterListener();
        // get current location to provide as defaults into
        // field
        LocationManager locationManager = (LocationManager) getActivity()
                .getSystemService(Context.LOCATION_SERVICE);
        // begin by getting the last known location
        Location fetchedLocationDetails = locationManager.getLastKnownLocation(provider);
        if (fetchedLocationDetails != null) {
            // update current location
            if (selectedFrame != null) {
                selectedFrame.setLocation(new be.hcpl.android.filmtag.model.Location(fetchedLocationDetails.getLatitude(), fetchedLocationDetails.getLongitude()));
                showLocation();
            }
        }
        // and start listening in order to update the location when more
        // information is retrieved
        // Register the listener with the Location Manager to receive location
        // updates
        locationManager
                .requestLocationUpdates(provider, 0, 0, locationListener);
    }

    /**
     * unregister location listeners
     */
    private void unregisterListener() {
        if (ContextCompat.checkSelfPermission(getActivity(),
                Manifest.permission.ACCESS_FINE_LOCATION) != PackageManager.PERMISSION_GRANTED) {
            return;
        }
        // get current location to provide as defaults into
        // field
        LocationManager locationManager = (LocationManager) getActivity()
                .getSystemService(Context.LOCATION_SERVICE);
        // remove previous listener first
        locationManager.removeUpdates(locationListener);
    }

    @Override
    public void onPause() {
        super.onPause();
        unregisterListener();
    }

    /**
     * listener for updating location when more data is found
     */
    private LocationListener locationListener = new LocationListener() {
        @Override
        public void onLocationChanged(final Location location) {
            // Called when a new location is found by the selected location
            // provider.
            if (selectedFrame != null) {
                selectedFrame.setLocation(new be.hcpl.android.filmtag.model.Location(location.getLatitude(), location.getLongitude()));
                // set on screen
                showLocation();
            }
        }

        @Override
        public void onStatusChanged(final String provider, final int status,
                                    final Bundle extras) {
            // nothing so far
        }

        @Override
        public void onProviderEnabled(final String provider) {
            // nothing so far
        }

        @Override
        public void onProviderDisabled(final String provider) {
            Toast.makeText(getActivity(), R.string.err_location_disabled, Toast.LENGTH_SHORT).show();
        }
    };

    private void showLocation() {
        if (selectedFrame != null && selectedFrame.getLocation() != null) {
            be.hcpl.android.filmtag.model.Location location = selectedFrame.getLocation();
            locationView.setText(getString(R.string.label_location) + " " + location.getLatitude() + " " + location.getLongitude());
        }
        markLocationAvailable();
    }

    private void updateItem() {
        // update values
        selectedFrame.setNotes(editNotes.getText().toString());

        try {
            selectedFrame.setAperture(Double.parseDouble(
                    getFieldTextOrDefault(editAperture, "key_default_apertures", String.valueOf(4))
            ));
        } catch (NumberFormatException nfe) {
            Toast.makeText(getActivity(), R.string.err_parsing_failed, Toast.LENGTH_SHORT).show();
        }

        try {
            selectedFrame.setShutter(Integer.parseInt(
                    getFieldTextOrDefault(editShutter, "key_default_shutter", String.valueOf(60))
            ));
        } catch (NumberFormatException nfe) {
            Toast.makeText(getActivity(), R.string.err_parsing_failed, Toast.LENGTH_SHORT).show();
        }
        selectedFrame.setLongExposure(checkLongExposure.isChecked());
        selectedFrame.setTags(Arrays.asList(TextUtils.split(editTags.getText().toString(), " ")));

        // store
        StorageUtil.updateFrames((MainActivity) getActivity(), roll, frames);

        // navigate back to overview
        backToOverview();
    }

    private String getFieldTextOrDefault(EditText field, String defaultKey, String hardDefault) {
        String text = field.getText().toString();
        if (text.equals("")) {
            SharedPreferences prefs = PreferenceManager.getDefaultSharedPreferences(getContext());
            return prefs.getString(defaultKey, hardDefault);
        }
        else {
            return text;
        }
    }

    private void backToOverview() {
        ((MainActivity) getActivity()).switchContent(FilmFrameListFragment.newInstance(roll));
    }

    @Override
    public void onResume() {
        super.onResume();
        ((MainActivity) getActivity()).setHomeAsUp(true);
        if (locationPermissionRequested) {
            getLocation();
            locationPermissionRequested = false;
        }
        if (storagePermissionRequested) {
            dispatchTakePictureIntent();
            storagePermissionRequested = false;
        }
        if( storagePermissionRequestedForPreview){
            loadImagePreview();
            storagePermissionRequestedForPreview = false;
        }
    }
}

File: app/src/main/java/be/hcpl/android/filmtag/EditFrameFragment.kt
package be.hcpl.android.filmtag

import android.Manifest
import android.app.Activity
import android.content.ActivityNotFoundException
import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.location.Location
import android.location.LocationListener
import android.location.LocationManager
import android.net.Uri
import android.os.Bundle
import android.os.Environment
import android.preference.PreferenceManager
import android.provider.MediaStore
import android.text.TextUtils
import android.view.Menu
import android.view.MenuInflater
import android.view.MenuItem
import android.view.View
import android.widget.CheckBox
import android.widget.DatePicker
import android.widget.EditText
import android.widget.ImageView
import android.widget.TextView
import android.widget.Toast
import androidx.core.content.ContextCompat
import androidx.core.content.ContextCompat.getDrawable
import androidx.core.os.bundleOf
import androidx.fragment.app.Fragment
import androidx.navigation.fragment.findNavController
import be.hcpl.android.filmtag.FilmFrameListFragment.Companion.KEY_FILM_ROLL
import be.hcpl.android.filmtag.model.Frame
import be.hcpl.android.filmtag.model.Roll
import be.hcpl.android.filmtag.util.StorageUtil
import com.google.android.material.datepicker.MaterialDatePicker
import java.io.File
import java.io.FileOutputStream
import java.io.IOException
import java.text.SimpleDateFormat
import java.util.Arrays
import java.util.Calendar
import java.util.Date
import java.util.Locale


/**
 * Created by hcpl on 1/08/15.
 */
class EditFrameFragment : Fragment(R.layout.fragment_form_frame) {

    // model object of a film or roll with a number of frames exposed at a given value
    private var roll: Roll? = null

    private var selectedFrame: Frame? = null
    private var previousFrame: Frame? = null

    private var frames: List<Frame>? = null

    // views
    private lateinit var editAperture: EditText
    private lateinit var editShutter: EditText
    private lateinit var editExposure: CheckBox
    private lateinit var editNotes: EditText
    private lateinit var editTags: EditText
    private lateinit var textNumber: TextView
    private lateinit var textLocation: TextView
    private lateinit var iconLocation: ImageView
    private lateinit var imagePreview: ImageView
    private lateinit var dateView: TextView

    // date picker
    private lateinit var datePicker: MaterialDatePicker<Long>
    val dateFormatter = SimpleDateFormat("yyyy-MM-dd", Locale.getDefault())

    private fun initDatePickerWith(date: Long) = MaterialDatePicker.Builder.datePicker()
        .setTitleText(getString(R.string.select_date))
        .setSelection(date)
        .build()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setHasOptionsMenu(true)
        restoreState(arguments)
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        outState.putSerializable(KEY_FRAMES, frames as ArrayList<*>?)
        outState.putInt(KEY_FRAME_IDX, frames!!.indexOf(selectedFrame))
        outState.putSerializable(KEY_ROLL, roll)
    }

    override fun onViewStateRestored(savedInstanceState: Bundle?) {
        super.onViewStateRestored(savedInstanceState)
        restoreState(savedInstanceState)
    }

    private fun restoreState(state: Bundle?) {
        state?.let {
            frames = state.getSerializable(KEY_FRAMES) as List<Frame>
            val selectedFrameIndex = state.getInt(KEY_FRAME_IDX)
            selectedFrame = frames!![selectedFrameIndex]
            previousFrame = if (selectedFrameIndex > 0) frames!![selectedFrameIndex - 1] else null
            roll = state.getSerializable(KEY_ROLL) as Roll
        }
    }

    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        inflater.inflate(R.menu.update_frame, menu)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        // find all views
        editAperture = view.findViewById(R.id.edit_aperture)
        editShutter = view.findViewById(R.id.edit_shutter)
        editExposure = view.findViewById(R.id.long_exposure)
        editNotes = view.findViewById(R.id.edit_notes)
        editTags = view.findViewById(R.id.edit_tags)
        iconLocation = view.findViewById(R.id.image_location_indicator)
        imagePreview = view.findViewById(R.id.image_preview)
        textLocation = view.findViewById(R.id.text_location)
        textNumber = view.findViewById(R.id.edit_number)
        dateView = view.findViewById(R.id.edit_date)

        // populate with frame details
        selectedFrame?.let {
            textNumber.text = "${it.number}"
            if (it.aperture != 0.0)
                editAperture.setText(it.aperture.toString())
            if (it.shutter != 0)
                editShutter.setText(it.shutter.toString())
            editExposure.isChecked = it.isLongExposure
            editNotes.setText(it.notes)
            // populate the tags here
            if (it.tags.isNotEmpty())
                editTags.setText(TextUtils.join(" ", it.tags))
            loadImagePreview() // load from frame storage
            showLocation()
            // update date
            val validDate = it.dateTaken?: Calendar.getInstance().timeInMillis
            dateView.setText(dateFormatter.format(validDate))
            datePicker = initDatePickerWith(validDate)
            dateView.setOnClickListener{
                datePicker.show(childFragmentManager, TAG_DATEPICKER)
            }
            datePicker.addOnPositiveButtonClickListener {
                // Respond to positive button click.
                dateView.text = dateFormatter.format(datePicker.selection)
            }
        }

        updateHints()

        // TODO implement autocomplete for inputs
    }

    // Uses previous frame values for aperture & shutter as hints if available
    // Otherwise uses default aperture & shutter from settings
    // Also, when previous frame has long-exposure checked, check it automatically
    private fun updateHints() {
        val prefs = PreferenceManager.getDefaultSharedPreferences(context)

        if (previousFrame != null && previousFrame!!.aperture != Frame.EMPTY_VALUE.toDouble()) {
            editAperture.hint = previousFrame!!.aperture.toString()
        } else {
            editAperture.hint = prefs.getString("key_default_apertures", 4.toString())
        }

        if (previousFrame != null && previousFrame!!.shutter != Frame.EMPTY_VALUE) {
            editShutter.hint = previousFrame!!.shutter.toString()
        } else {
            editShutter.hint = prefs.getString("key_default_shutter", 60.toString())
        }

        if (previousFrame != null && previousFrame!!.isLongExposure) {
            editExposure.isChecked = true
        }
    }

    private fun updateLocationViews() {
        selectedFrame?.location?.let {
            iconLocation.setImageDrawable(
                getDrawable(requireContext(), R.drawable.ic_action_device_gps_primary)
            )
            textLocation.setOnClickListener {
                showOnMap(Uri.parse("geo: ${selectedFrame?.location?.latitude},${selectedFrame?.location?.longitude}"))
            }
        } ?: {
            // no known location set
            iconLocation.setImageDrawable(
                getDrawable(requireContext(), R.drawable.ic_action_device_gps_silver)
            )
            textLocation.setOnClickListener { getLocation() }
        }
    }

    private fun showOnMap(geoLocation: Uri?) {
        geoLocation?.let {
            val intent = Intent(Intent.ACTION_VIEW)
            intent.data = geoLocation
            startActivity(intent)
        }
    }

    private fun loadImagePreview() {
        // and if path set try loading after permission check
        if (selectedFrame?.pathToImage != null) {
            try {
                val options = BitmapFactory.Options()
                options.inSampleSize = 4
                options.inJustDecodeBounds = false
                val bm = BitmapFactory.decodeFile(selectedFrame?.pathToImage, options)
                imagePreview.setImageBitmap(bm)
            } catch (e: Exception) {
                // ignore any exceptions here
                handleImageError()
            }
        }
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.action_update -> {
                updateItem()
                return true
            }

            android.R.id.home -> {
                backToDetail()
                return true
            }

            R.id.action_camera -> {
                dispatchTakePictureIntent()
                return true
            }

            R.id.action_location -> {
                getLocation()
                return true
            }
        }
        return false
    }

    private fun dispatchTakePictureIntent() {
        val takePictureIntent = Intent(MediaStore.ACTION_IMAGE_CAPTURE)
        try {
            startActivityForResult(takePictureIntent, REQUEST_IMAGE_CAPTURE)
        } catch (e: ActivityNotFoundException) {
            // display error state to the user
            handleImageError();
        }
    }

    private fun handleImageError() {
        Toast.makeText(requireContext(), R.string.error_creating_image, Toast.LENGTH_LONG).show()
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        if (requestCode == REQUEST_IMAGE_CAPTURE && resultCode == Activity.RESULT_OK) {
            loadImagePreviewFromIntent(data)
        }
    }

    private fun loadImagePreviewFromIntent(data: Intent?) {
        val imageBitmap = data?.extras?.get("data") as Bitmap
        imagePreview.setImageBitmap(imageBitmap);
        val file = createImageFile()
        try {
            FileOutputStream(file.absolutePath).use { out ->
                imageBitmap.compress(Bitmap.CompressFormat.JPEG, 100, out)
            }
            // save image path at this point
            StorageUtil.updateFrames(activity as MainActivity, roll!!, frames!!)
        } catch (e: IOException) {
            handleImageError()
        }
    }

    @Throws(IOException::class)
    private fun createImageFile(): File {
        // Create an image file name
        val timeStamp = SimpleDateFormat("yyyyMMdd_HHmmss").format(Date())
        val imageFileName = "JPEG_" + timeStamp + "_"
        val storageDir = Environment.getExternalStoragePublicDirectory(
            Environment.DIRECTORY_PICTURES
        )
        val image = File.createTempFile(
            imageFileName, /* prefix */
            ".jpg", /* suffix */
            storageDir      /* directory */
        )

        // Save a file: path for use with ACTION_VIEW intents
        selectedFrame!!.pathToImage = image.absolutePath
        return image
    }

    private fun getLocation() {
        if (ContextCompat.checkSelfPermission(
                requireContext(),
                Manifest.permission.ACCESS_FINE_LOCATION
            ) != PackageManager.PERMISSION_GRANTED
        ) {

            // No explanation needed, we can request the permission.
            if (!locationPermissionRequested) {
                locationPermissionRequested = true
                requestPermissions(
                    arrayOf(Manifest.permission.ACCESS_FINE_LOCATION),
                    MY_PERMISSIONS_REQUEST_LOCATION
                )
            }
            return
        }
        registerLocationListener(LocationManager.GPS_PROVIDER)
    }

    private fun registerLocationListener(provider: String) {
        if (ContextCompat.checkSelfPermission(
                requireContext(),
                Manifest.permission.ACCESS_FINE_LOCATION
            ) != PackageManager.PERMISSION_GRANTED
        ) {
            return
        }
        // remove previous listener first
        unregisterListener()
        // get current location to provide as defaults into
        // field
        val locationManager =
            requireContext().getSystemService(Context.LOCATION_SERVICE) as LocationManager
        // begin by getting the last known location
        val fetchedLocationDetails = locationManager.getLastKnownLocation(provider)
        if (fetchedLocationDetails != null) {
            // update current location
            if (selectedFrame != null) {
                selectedFrame!!.location = be.hcpl.android.filmtag.model.Location(
                    fetchedLocationDetails.latitude,
                    fetchedLocationDetails.longitude
                )
                showLocation()
            }
        }
        // and start listening in order to update the location when more
        // information is retrieved
        // Register the listener with the Location Manager to receive location
        // updates
        locationManager
            .requestLocationUpdates(provider, 0, 0f, locationListener)
    }

    private fun unregisterListener() {
        if (ContextCompat.checkSelfPermission(
                requireContext(),
                Manifest.permission.ACCESS_FINE_LOCATION
            ) != PackageManager.PERMISSION_GRANTED
        ) {
            return
        }
        // get current location to provide as defaults into
        // field
        val locationManager =
            requireContext().getSystemService(Context.LOCATION_SERVICE) as LocationManager
        // remove previous listener first
        locationManager.removeUpdates(locationListener)
    }

    override fun onPause() {
        super.onPause()
        unregisterListener()
    }

    /**
     * listener for updating location when more data is found
     */
    private val locationListener = object : LocationListener {
        override fun onLocationChanged(location: Location) {
            // Called when a new location is found by the selected location
            // provider.
            if (selectedFrame != null) {
                selectedFrame!!.location =
                    be.hcpl.android.filmtag.model.Location(location.latitude, location.longitude)
                // set on screen
                showLocation()
            }
        }

        override fun onStatusChanged(
            provider: String, status: Int,
            extras: Bundle
        ) {
            // nothing so far
        }

        override fun onProviderEnabled(provider: String) {
            // nothing so far
        }

        override fun onProviderDisabled(provider: String) {
            Toast.makeText(activity, R.string.err_location_disabled, Toast.LENGTH_SHORT).show()
        }
    }

    private fun showLocation() {
        selectedFrame?.location?.let { location ->
            textLocation.text =
                "${getString(R.string.label_location)} ${location.latitude} ${location.longitude}"
            updateItem(false) // also update location in storage autom.
        }
        updateLocationViews()
    }

    private fun updateItem(navigateBack: Boolean = true) {
        // update values
        selectedFrame!!.notes = editNotes.text.toString()

        try {
            selectedFrame!!.aperture =
                java.lang.Double.parseDouble(getFieldTextOrHint(editAperture))
        } catch (_: NumberFormatException) {
            Toast.makeText(activity, R.string.err_parsing_failed, Toast.LENGTH_SHORT).show()
        }

        try {
            selectedFrame!!.shutter = Integer.parseInt(getFieldTextOrHint(editShutter))
        } catch (_: NumberFormatException) {
            Toast.makeText(activity, R.string.err_parsing_failed, Toast.LENGTH_SHORT).show()
        }
        try {
            selectedFrame!!.dateTaken = dateView.text?.let { dateFormatter.parse(it.toString()).time }
        } catch (_: NumberFormatException) {
            //fail silently for date
            // Toast.makeText(activity, R.string.err_parsing_failed, Toast.LENGTH_SHORT).show()
        }
        selectedFrame!!.isLongExposure = editExposure.isChecked
        selectedFrame!!.tags = Arrays.asList(*TextUtils.split(editTags.text.toString(), " "))

        // store
        StorageUtil.updateFrames(activity as MainActivity, roll!!, frames!!)

        // navigate back to overview
        if (navigateBack) backToDetail()
    }

    private fun getFieldTextOrHint(field: EditText): String {
        val text = field.text.toString()
        return if (TextUtils.isEmpty(text)) {
            field.hint.toString()
        } else {
            text
        }
    }

    private fun backToDetail() {
        findNavController().navigate(R.id.action_detail, bundleOf(KEY_FILM_ROLL to roll))
    }

    override fun onResume() {
        super.onResume()
        if (locationPermissionRequested) {
            getLocation()
            locationPermissionRequested = false
        }
    }

    companion object {

        // all image related code from http://developer.android.com/training/camera/photobasics.html

        const val KEY_FRAME_IDX = "frame_index"
        const val KEY_FRAMES = "frames"
        const val KEY_ROLL = "roll"
        const val TAG_DATEPICKER = "datePicker"

        private val MY_PERMISSIONS_REQUEST_LOCATION = 100
        private val MY_PERMISSIONS_REQUEST_STORAGE = 200
        private val REQUEST_IMAGE_CAPTURE = 300

        private var locationPermissionRequested = false
        private var storagePermissionRequested = false
        private var storagePermissionRequestedForPreview = false

        fun newInstance(roll: Roll?, frames: List<Frame>?, frame: Int): EditFrameFragment {
            val args = Bundle()
            args.putSerializable(KEY_FRAMES, frames as ArrayList<*>)
            args.putInt(KEY_FRAME_IDX, frame)
            args.putSerializable(KEY_ROLL, roll)
            val fragment = EditFrameFragment()
            fragment.arguments = args
            return fragment
        }
    }
}
